### PR TITLE
Schedule deploy stage

### PR DESCRIPTION
**Pull request by anwar6953**
Adding the schedule from the AppConfigurer to the ProgramScheduleStoreDataset.

https://builds.cask.co/browse/CDAP-DUT5697-1

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/app/AbstractApplication.java [hunk 1]
```diff
@@ -170,4 +173,16 @@ protected void scheduleWorkflow(Schedule schedule, String workflowName) {
   protected void scheduleWorkflow(Schedule schedule, String workflowName, Map<String, String> properties) {
     configurer.addSchedule(schedule, SchedulableProgramType.WORKFLOW, workflowName, properties);
   }
+
+
+  /**
+   * Schedule the specified {@link Workflow}
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/app/AbstractApplication.java:177]: This method actually doesn't schedule a workflow, right? This method is to return a `ScheduleConfigurer` and the workflow will only get scheduled when one of the `trigger` method is called, right? Maybe better calling it `configureSchedule`?? Also according to `SchedulableProgramType`, seems like not just workflow can be schedule? Although that class says `CUSTOM_ACTION` can be scheduled, which doesn't really make sense. Can you verify what that enum is for?

BTW, if we want to extend that to be able to schedule other programs, do we simply just add new methods?

File: cdap-api/src/main/java/co/cask/cdap/api/app/ApplicationConfigurer.java [hunk 2]
```diff
@@ -100,4 +101,13 @@
...
                    Map<String, String> properties);
+
+  /**
+   * Schedule the specified {@link Workflow}
+   * @param scheduleName the name of the schedule
+   * @param workflowName the name of the Workflow
+   *
+   * @return The {@link ScheduleConfigurer} used to configure the schedule
+   */
+  ScheduleConfigurer scheduleWorkflow(String scheduleName, String workflowName);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/app/ApplicationConfigurer.java]: Same as above.

File: cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/ConcurrencyConstraint.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.schedule.constraint;
+
+/**
+ * A constraint which dictates an upper bound on the number of concurrent schedule runs.
+ */
+public class ConcurrencyConstraint extends Constraint {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/ConcurrencyConstraint.java]: So you have two different set of Constraint classes? One set in API, the other in app-fabric? Maybe better to call the one in api differently? E.g. `ScheduleConstraint` and `ScheduleTrigger`?

File: cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/TimeRangeConstraint.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * A Constraint that defines a time range in which the schedule is allowed to execute.
+ */
+public class TimeRangeConstraint extends Constraint {
+
+  // only is satisfied within the range [startHour, endHour)
+  // TODO: Allow minute granularity
+  private final int startHour;
+  private final int endHour;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/TimeRangeConstraint.java]: Need the timezone as well? Actually need to indicate does it stick with a particular timezone and using the local timezone. This is mainly for handling cross data center replication case.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java [hunk 5]
```diff
@@ -194,6 +199,13 @@ public void addSchedule(Schedule schedule, SchedulableProgramType programType, S
     schedules.put(schedule.getName(), spec);
   }
 
+  @Override
+  public ScheduleConfigurer scheduleWorkflow(String scheduleName, String workflowName) {
+    AtomicReference<ScheduleCreationSpec> reference = new AtomicReference<>();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java]: Seems a bit overly complicated. You can just pass the `programSchedules` map to the `DefaultScheduleConfigurer`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/TriggerCodec.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  // we use a separate GSON object to avoid recursion
+  private static final Gson GSON = new Gson();
+
+  @Nullable
+  private JsonElement serializeTrigger(@Nullable Trigger trigger,
+                                       JsonSerializationContext jsonSerializationContext) {
+    if (trigger == null) {
+      return null;
+    }
+    JsonArray jsonArray = new JsonArray();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/TriggerCodec.java]: Prefer to use `JsonObject` instead of array so that we have keys to identify the fields to allow future mutation easier.

**Review by chtyim — COMMENTED** [review]
Some high level comments. Will look more closely.

File: cdap-api/src/main/java/co/cask/cdap/api/app/AbstractApplication.java (hunk 1 above)
Comment (anwar6953) [reply cdap-api/src/main/java/co/cask/cdap/api/app/AbstractApplication.java:177]: Only workflow can be scheduled. That enum is for what the program types are allowed to be in a workflow (its a bit confusingly-named).

File: cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/ConcurrencyConstraint.java (hunk 3 above)
Comment (anwar6953) [reply cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/ConcurrencyConstraint.java]: Removed the ones from cdap-app-fabric for now. We can decide how to have it when we implement https://issues.cask.co/browse/CDAP-11338

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/internal/schedule/constraint/ConcurrencyConstraint.java]: In that case, I would suggest only have the abstract Constraint class in the cdap-api and the implementations inside cdap-app-fabric. Same goes for `Trigger`.

File: cdap-api/src/main/java/co/cask/cdap/api/app/AbstractApplication.java (hunk 1 above)
Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/app/AbstractApplication.java:177]: Do you have an example of what this will look like in application code? I feel that calling it addSchedule() would make sense... Perhaps we can have something like:
```
  addSchedule("mySchedule")
    .forWorkflow("wf1")
    .setDescription("my schedule")
    .addConstraint(...)
    .triggerOn(...);
```

Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/app/AbstractApplication.java:177]: and in the future we could simply add methods like forMapReduce() or forQuery()

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateProgramSchedulesStage.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    if (!input.canUpdateSchedules()) {
+      // if we cant update schedules, emit and return
+      emit(input);
+      return;
+    }
+
+    Map<String, ScheduleCreationSpec> programSchedules = input.getSpecification().getProgramSchedules();
+    for (final Map.Entry<String, ScheduleCreationSpec> entry : programSchedules.entrySet()) {
+
+          ScheduleCreationSpec scheduleCreationSpec = entry.getValue();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateProgramSchedulesStage.java]: Seems like the alignment is off.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java [hunk 8]
```diff
@@ -51,7 +51,7 @@ public CreateSchedulesStage(Scheduler scheduler) {
   }
 
   @Override
-  public void process(ApplicationWithPrograms input) throws Exception {
+  public void process(final ApplicationWithPrograms input) throws Exception {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: unnecessary change

**Review by chtyim — COMMENTED** [review]
Just two simple comment left. Rest LGTM
